tensorflow/compiler/mlir/tosa/transforms/legalize_tfl.cc

```diff
@@ -387,6 +388,57 @@ LogicalResult ConvertTFLRelu1Op::matchAndRewrite(
   return success();
 }
 
+LogicalResult ConvertTFLRelu0To1Op::matchAndRewrite(
+    Operation* op, PatternRewriter& rewriter) const {
+  auto tfl_relu0to1_op = cast<TFL::Relu0To1Op>(op);
+
+  ShapedType input_type = tfl_relu0to1_op.x().getType().cast<ShapedType>();
+  ShapedType output_type =
+      tfl_relu0to1_op.getResult().getType().cast<ShapedType>();
+
+  bool input_is_qtype =
+      input_type.getElementType().isa<mlir::quant::UniformQuantizedType>();
+  bool output_is_qtype =
+      output_type.getElementType().isa<mlir::quant::UniformQuantizedType>();
+
+  if (input_is_qtype != output_is_qtype) {
+    return rewriter.notifyMatchFailure(op,
+        "input/output tensor should be all quantized or all floating-point");
+  }
+
+  int64_t clamp_min = 0;
+  int64_t clamp_max = 1;
+  Value clamp_in = tfl_relu0to1_op.x();
+
+  if (output_is_qtype && input_is_qtype) {
+    UniformQuantizedType input_qtype =
+        input_type.getElementType()
+            .cast<mlir::quant::UniformQuantizedType>();
+    UniformQuantizedType output_qtype =
+        output_type.getElementType()
+            .cast<mlir::quant::UniformQuantizedType>();
+
+    clamp_min = output_qtype.getZeroPoint();
+
+    clamp_max = std::llround(1.0f / output_qtype.getScale()) +
```

Comment:
Remember to  sent the clamp min/max within the bounds of the type. It is possible that 0 or 1 (or both) are outside the bounds of the representation.

Reply:
I'm not too sure if I understood that correctly. Here the type for clamp_min/max is int64_t where 0 or 1 are impossible to get outside of the range?

Reply:
So you are clamping to the quantized equivalents of 0.0 and 1.0 in an intN_t value. However it is possible that the quantized equivalent values are outside the bounds of the int8_t we use to represent. So each should be clamped to the limits of those bounds before constructing the rescale value.

E.g. it is nonsensical to clamp an int8_t to -[2321 12321], and instead we should clamp to [-128 127].

Reply:
Or are we able to guarantee the clamping occurs within a 64-bit value?

Reply:
I am not too sure about the guarantee. I simply copy-pasted the code from other Relu legalizations. Maybe I can do another patch to fix everybody?

Reply:
That would definitely be appreciated. The TosaToLinalg does perform additional clamping on the values but defense-in-depth feels like the right choice.

We should also consider having a verifier for `tosa.clamp` upstreamed that verifies the clamp is reasonable. It can guarantee non-linalg backends don't need to check that the bounds are reasonable.

Reply:
Yup, I saw that: https://github.com/llvm/llvm-project/blob/main/mlir/lib/Conversion/TosaToLinalg/TosaToLinalg.cpp

I'm planning on just leave it as it is right now for Relu0To1 and do a full update patch for all the members in the Relu familiy on this min/max check. Sounds good?

Reply:
Or do a verifier to `tosa.clamp`

Reply:
Yup you can do it in a follow up but both clamping and verifying would be good.

Reply:
Sounds good. I will add that to my next Q's plan.